services/venice-controller/src/main/java/com/linkedin/venice/controller/kafka/protocol/serializer/AdminOperationSerializer.java

```diff
@@ -5,50 +5,81 @@
 import org.apache.avro.io.DecoderFactory;
 import org.apache.avro.io.Encoder;
 import org.apache.avro.specific.SpecificDatumReader;
-import org.apache.avro.specific.SpecificDatumWriter;
 
 
 public class AdminOperationSerializer {
   // Latest schema id, and it needs to be updated whenever we add a new version
   public static final int LATEST_SCHEMA_ID_FOR_ADMIN_OPERATION =
       AvroProtocolDefinition.ADMIN_OPERATION.getCurrentProtocolVersion();
 
-  private static SpecificDatumWriter<AdminOperation> SPECIFIC_DATUM_WRITER =
-      new SpecificDatumWriter<>(AdminOperation.getClassSchema());
+  public static final Schema LATEST_SCHEMA = AdminOperation.getClassSchema();
+
   /** Used to generate decoders. */
   private static final DecoderFactory DECODER_FACTORY = new DecoderFactory();
 
   private static final Map<Integer, Schema> PROTOCOL_MAP = initProtocolMap();
 
-  public byte[] serialize(AdminOperation object) {
-    try {
-      ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();
-      Encoder encoder = AvroCompatibilityHelper.newBinaryEncoder(byteArrayOutputStream, true, null);
-      SPECIFIC_DATUM_WRITER.write(object, encoder);
-      encoder.flush();
+  /**
+   * Serialize AdminOperation object to bytes.
+   *
+   * <p>
+   * If writerSchemaId equals LATEST_SCHEMA_ID_FOR_ADMIN_OPERATION, return the bytes[] from the first serialization.
+   * Otherwise, serialize the object to the writer schema (lower version).
+   * </p>
+   *
+   * <p>
+   * This involves:
+   * <ol>
+   *   <li>Serializing the object to a GenericRecord with the latest schema.</li>
+   *   <li>Deserializing it to a GenericRecord with the writer schema.</li>
+   *   <li>Serializing it to bytes.</li>
+   * </ol>
+   * </p>
+   * <p>
+   * This process ensures the object is serialized to the lower schema version.
+   * The normal serialization process may fail (ClassCastException) due to:
+   * <ul>
+   *   <li>Differences in field types</li>
+   *   <li>New fields added in the middle of the schema instead of at the end</li>
+   * </ul>
+   * </p>
+   */
+  public byte[] serialize(AdminOperation object, int writerSchemaId) {
+    byte[] serializedBytes = serialize(object, LATEST_SCHEMA);
 
-      return byteArrayOutputStream.toByteArray();
-    } catch (IOException e) {
-      throw new VeniceMessageException("Failed to encode message: " + object.toString(), e);
+    // If writerSchema is the latest schema, we can return the serialized bytes directly.
+    if (writerSchemaId == LATEST_SCHEMA_ID_FOR_ADMIN_OPERATION) {
+      return serializedBytes;
     }
+
+    // Get the writer schema.
+    Schema writerSchema = getSchema(writerSchemaId);
+
+    // If writerSchema is not the latest schema, we need to deserialize the serialized bytes to GenericRecord with
+    // the writer schema, then serialize it to bytes with the writer schema.
+    GenericRecord genericRecord = deserialize(serializedBytes, LATEST_SCHEMA, writerSchema);
```

Comment:
We used writerSchema on the readerSchema argument position of the deserialize API, which is correct but could be confusing.
Talked to @minhmo1620 ; we are going to change it to `targetSchema`

Reply:
Updated. Please help check the new signature for serializer/deserializer.